src/trace_ng/log.zig

```diff
@@ -0,0 +1,672 @@
+                level,
+                maybe_scope,
+                message,
+                null,
+                null,
+                null,
+            );
+            self.channel.send(log_message) catch @panic("could not send to channel");
+        }
+
+        pub fn logWithFields(self: *Self, level: Level, message: []const u8, fields: anytype) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+            const maybe_scope = blk: {
+                if (scope) |s| {
+                    break :blk @typeName(s);
+                } else {
+                    break :blk null;
+                }
+            };
+
+            const log_message = logfmt.createLogMessage(
+                &self.recycle_fba,
+                self.max_buffer,
+                level,
+                maybe_scope,
+                message,
+                fields,
+                null,
+                null,
+            );
+            self.channel.send(log_message) catch @panic("could not send to channel");
+        }
+
+        pub fn logf(self: *Self, level: Level, comptime fmt: []const u8, args: anytype) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+            const maybe_scope = blk: {
+                if (scope) |s| {
+                    break :blk @typeName(s);
+                } else {
+                    break :blk null;
+                }
+            };
+
+            const log_message = logfmt.createLogMessage(
+                &self.recycle_fba,
+                self.max_buffer,
+                level,
+                maybe_scope,
+                null,
+                null,
+                fmt,
+                args,
+            );
+            self.channel.send(log_message) catch @panic("could not send to channel");
+        }
+
+        pub fn logfWithFields(self: *Self, level: Level, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+            const maybe_scope = blk: {
+                if (scope) |s| {
+                    break :blk @typeName(s);
+                } else {
+                    break :blk null;
+                }
+            };
+
+            const log_message = logfmt.createLogMessage(
+                &self.recycle_fba,
+                self.max_buffer,
+                level,
+                maybe_scope,
+                null,
+                fields,
+                fmt,
+                args,
+            );
+            self.channel.send(log_message) catch @panic("could not send to channel");
+        }
+    };
+
+    const TestingLogger = struct {
+        const Self = @This();
+        max_level: Level,
+        allocator: Allocator,
+        recycle_fba: RecycleFBA,
```

Comment:
You don't need an allocator in this logger. You can format strings on the stack since they're not being sent over a channel